### PR TITLE
feat: [Custom widget] Enable dynamic height and auto layout.

### LINKED ISSUE #30900: [enhancement]: [custom widgets] dynamic height and auto layout configuration

### PR DESCRIPTION
**Pull request by sbalaji1192**
Now custom widgets can grow in height in fixed (dynamic height) and in auto layout. This will enable users to implement expand/collapse , Table with expandable rows and more

#### PR fixes following issue(s)
Fixes #30900

#### Type of change

- New feature (non-breaking change which adds functionality)
>
>
>
- [ ] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **New Features**
	- Introduced dynamic height adjustment for custom widgets, enhancing the flexibility and visual integration within the app.
- **Improvements**
	- Optimized height update mechanism for custom widgets using `ResizeObserver` for more responsive layout changes.
- **Refactor**
	- Streamlined event handling and UI adjustments in custom widgets to support dynamic height settings more efficiently.

### CODE REVIEW

**Review by somangshu — APPROVED** [review]
Code LGTM

**Comment by somangshu** [comment]
https://github.com/appsmithorg/appsmith/assets/11089579/20edc51b-9a69-4f11-b067-fd1802385ea9

On the default mode, if I change to auto-height, I see an un-expected result.

**Comment by somangshu** [comment]
https://github.com/appsmithorg/appsmith/assets/11089579/68fe0cea-455c-47ec-856b-e5d9ba891d9b

related to above, on auto height with limits, I can't reduce the threshold below the bounding box, if I extend it further, the custom widget grows until that point.

**Comment by somangshu** [comment]
@sbalaji1192 some issues noticed on the PR.

**Comment by sbalaji1192** [comment]
@somangshu Fixed the issue. Please try again.

File: app/client/src/widgets/CustomWidget/widget/index.tsx [hunk 1]
```diff
@@ -85,6 +86,27 @@ class CustomWidget extends BaseWidget<CustomWidgetProps, WidgetState> {
         sectionIndex: 2,
         active: true,
         defaultValue: DynamicHeight.FIXED,
+        helperText: (props: WidgetProps) => {
+          if (props.dynamicHeight !== DynamicHeight.FIXED) {
+            return (
+              <div className="leading-5 mt-[10px]">
```
Comment (somangshu) [inline app/client/src/widgets/CustomWidget/widget/index.tsx:92]: Shall we change the message to:
`For the auto-height feature to function correctly, the custom widget's container should not have a fixed height set. Read more`

WDYT?

Comment (sbalaji1192) [reply app/client/src/widgets/CustomWidget/widget/index.tsx:92]: Sure Somangshu. Let me change

**Review by somangshu — APPROVED** [review]
Changes LGTM

**Review by somangshu — APPROVED** [review]
LGTM

**Comment by somangshu** [comment]
@sbalaji1192 lets plan for the change we are going to make in the docs, would be great if we can show it with examples. 

cc @harshilp24

**Pull request merged**